test/functional/lua/diagnostic_spec.lua

```diff
@@ -0,0 +1,795 @@
+local helpers = require('test.functional.helpers')(after_each)
+
+local clear = helpers.clear
+local exec_lua = helpers.exec_lua
+local eq = helpers.eq
+local nvim = helpers.nvim
+
+describe('vim.diagnostic', function()
+  before_each(function()
+    clear()
+
+    exec_lua [[
+      require('vim.diagnostic')
+
+      make_error = function(msg, x1, y1, x2, y2)
+        return {
+          lnum = x1,
+          col = y1,
+          end_lnum = x2,
+          end_col = y2,
+          message = msg,
+          severity = vim.diagnostic.severity.ERROR,
+        }
+      end
+
+      make_warning = function(msg, x1, y1, x2, y2)
+        return {
+          lnum = x1,
+          col = y1,
+          end_lnum = x2,
+          end_col = y2,
+          message = msg,
+          severity = vim.diagnostic.severity.WARN,
+        }
+      end
+
+      make_information = function(msg, x1, y1, x2, y2)
+        return {
+          lnum = x1,
+          col = y1,
+          end_lnum = x2,
+          end_col = y2,
+          message = msg,
+          severity = vim.diagnostic.severity.INFO,
+        }
+      end
+
+      count_extmarks = function(namespace, bufnr)
+        return #vim.api.nvim_buf_get_extmarks(bufnr, namespace, 0, -1, {})
+      end
+    ]]
+
+    exec_lua([[
+      diagnostic_ns = vim.api.nvim_create_namespace("diagnostic_spec")
+      other_ns = vim.api.nvim_create_namespace("other_namespace")
+      diagnostic_bufnr = vim.api.nvim_create_buf(true, false)
+      local lines = {"1st line of text", "2nd line of text", "wow", "cool", "more", "lines"}
+      vim.fn.bufload(diagnostic_bufnr)
+      vim.api.nvim_buf_set_lines(diagnostic_bufnr, 0, 1, false, lines)
+      return diagnostic_bufnr
+    ]])
+  end)
+
+  after_each(function()
+    clear()
+  end)
+
+  it('creates highlight groups', function()
+    eq({
+      'DiagnosticDefaultError',
+      'DiagnosticDefaultHint',
+      'DiagnosticDefaultInfo',
+      'DiagnosticDefaultWarn',
+      'DiagnosticFloatingError',
+      'DiagnosticFloatingHint',
+      'DiagnosticFloatingInfo',
+      'DiagnosticFloatingWarn',
+      'DiagnosticSignError',
+      'DiagnosticSignHint',
+      'DiagnosticSignInfo',
+      'DiagnosticSignWarn',
+      'DiagnosticUnderlineError',
+      'DiagnosticUnderlineHint',
+      'DiagnosticUnderlineInfo',
+      'DiagnosticUnderlineWarn',
+      'DiagnosticVirtualTextError',
+      'DiagnosticVirtualTextHint',
+      'DiagnosticVirtualTextInfo',
+      'DiagnosticVirtualTextWarn',
+    }, exec_lua([[return vim.fn.getcompletion('Diagnostic', 'highlight')]]))
+  end)
+
+  describe('vim.diagnostic', function()
+    it('should be able to retrieve diagnostics from all buffers and namespaces', function()
```

Comment:
just mentioning this to avoid proliferation: "should be able to" is noise in test descriptions.

same for the other cases in this PR.
```suggestion
    it('retrieves diagnostics from all buffers and namespaces', function()
```